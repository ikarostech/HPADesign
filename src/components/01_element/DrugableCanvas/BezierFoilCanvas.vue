<template>
  <canvas
    ref="drawCanvas"
    :width="width"
    :height="height"
    v-on:mousedown="mousedown"
    v-on:mouseup="mouseup"
    v-on:mousemove="mousemove"
  />
</template>

<script lang="ts">
import assert from "assert";
import { LinearInterpolite } from "@/model/math/LinearInterpolite/LinearInterpolite";
import { Vector2 } from "three/src/math/Vector2";
import Vue, { PropType } from "vue";
import { BezierControls } from "./BezierControls";
import { CanvasPoint } from "./CanvasPoint";

//import { Vector2 } from 'three/src/math/Vector2';

export function buildBezierControls(
  controls: BezierControls,
  linspaceSize: number
): Vector2[] {
  const bezierCurve: Vector2[] = [];

  //TODO BezierPointの仕様変更
  for (let i = 0; i < controls.points.length - 1; i += 2) {
    const p1: Vector2 = controls.points[i].point;
    const c: Vector2 = controls.points[i + 1].point;
    const p2: Vector2 = controls.points[i + 2].point;

    for (let n = 0; n < linspaceSize; n++) {
      const t: number = n / linspaceSize;

      const pos: Vector2 = p1
        .clone()
        .multiplyScalar(Math.pow(1 - t, 2))
        .add(c.clone().multiplyScalar(2 * t * (1 - t)))
        .add(p2.clone().multiplyScalar(Math.pow(t, 2)));

      bezierCurve.push(pos);
    }
  }
  bezierCurve.push(new Vector2(1, 0));
  return bezierCurve;
}

export type select = {
  isTop: boolean;
  index: number;
}
export type DataType = {
  topControls: BezierControls;
  canverControls: BezierControls;
  canvas: any;
  ctx: any;
  width: number;
  height: number;
	isDrag: boolean;
	selected: select | null;
};

export default Vue.extend({
  name: "App",
  props: {
    radius: Number,
  },
  data: function (): DataType {
    const topControls = new BezierControls();
    topControls.points = [
      {
            point: new Vector2(0.0, 0.0),
            visible: false,
          },
          {
            point: new Vector2(0.0, 0.09),
            visible: true,
          },
          {
            point: new Vector2(0.34, 0.09),
            visible: true,
          },
          {
            point: new Vector2(0.6, 0.09),
            visible: true,
          },
          {
            point: new Vector2(1.0, 0.0),
            visible: false,
          },
    ];
    const canverControls = new BezierControls();
    canverControls.points = [
          {
            point: new Vector2(0.0, 0.0),
            visible: false,
          },
          {
            point: new Vector2(0.2, 0.03),
            visible: true,
          },
          {
            point: new Vector2(0.55, 0.03),
            visible: true,
          },
          {
            point: new Vector2(0.65, 0.03),
            visible: true,
          },
          {
            point: new Vector2(1, 0),
            visible: false,
          },
    ];
    return {
      topControls: topControls,
      canverControls: canverControls,
      canvas: null,
      ctx: null,
      width: 600,
      height: 200,
			isDrag: false,
      selected: null,
    };
  },

  mounted: function () {
    this.canvas = this.$refs.drawCanvas;
    this.ctx = this.canvas.getContext("2d");
    this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
    this.draw();
  },
  computed: {
    getShape: function (): Vector2[] {
      //翼型上面
      const topShape: Vector2[] = LinearInterpolite(
        buildBezierControls(this.topControls, 100),
        81
      );
      const canverShape: Vector2[] = LinearInterpolite(
        buildBezierControls(this.canverControls, 100),
        81
      );
      assert(topShape.length == canverShape.length);

      const bottomShape: Vector2[] = [];
      for (let i = 1; i < topShape.length; i++) {
        const element: Vector2 = canverShape[i]
          .multiplyScalar(2)
          .sub(topShape[i]);
        bottomShape.push(element);
      }
      assert(topShape.length == bottomShape.length + 1);

      const shape = topShape.reverse().concat(bottomShape);
      this.$emit("updateShape", shape);
      return shape;
    },
  },

  methods: {
    mousedown: function (event: {target: HTMLCanvasElement; clientX: number; clientY: number }): void {
			this.isDrag = true;
			this.selected = null;

			const rect = event.target.getBoundingClientRect();
      const mousePos = new Vector2(event.clientX - rect.left - 100, -(event.clientY - rect.top - 100)).divideScalar(400);

      const topControlFirstMatch: number = this.topControls.points
        .findIndex(controlPoint => controlPoint.visible === true && controlPoint.point.distanceTo(mousePos) <= this.radius / 400)
      
      if (topControlFirstMatch >= 0) {
        this.selected = {
          isTop: true,
          index: topControlFirstMatch
        };
        return;
      }

      const canverControlFirstMatch: number = this.canverControls.points
        .findIndex(controlPoint => controlPoint.visible === true && controlPoint.point.distanceTo(mousePos) <= this.radius / 400)
      
      if (canverControlFirstMatch >= 0) {
        this.selected = {
          isTop: false,
          index: canverControlFirstMatch
        };
        return;
      }		
      
			return;
    },
    mouseup: function (): void {
      this.isDrag = false;
      this.selected = null;
    },
    mousemove: function (event: {target: HTMLCanvasElement; clientX: number; clientY: number }): void {
      
			if (this.selected === null || this.isDrag === false) {
				return;
      }
      
			const rect = event.target.getBoundingClientRect();
      const mousePos = new Vector2(event.clientX - rect.left - 100, -(event.clientY - rect.top - 100)).divideScalar(400);
      if (this.selected.isTop) {
        this.topControls.points[this.selected.index].point = mousePos.clone();
      } 
      else {
        this.canverControls.points[this.selected.index].point = mousePos.clone();
      }

      this.draw();
		},
    draw: function (): void {
      this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height)
			this.drawCircle();
			this.drawLine();
    },
    drawCircle: function (): void {
      const curves = [this.topControls, this.canverControls];
      curves.forEach((curve) => {
        curve.points
          .filter((item) => item.visible)
          .map((item) => item.point)
          .forEach((point) => {
            this.ctx.beginPath();
            this.ctx.arc(
              point.x * 400 + 100,
              -point.y * 400 + 100,
              this.radius,
              0,
              Math.PI * 2,
              false
            );
            this.ctx.fillStyle = curve.controlPointStyle;
            this.ctx.fill();
            this.ctx.strokeStyle = curve.controlPointStyle;
            this.ctx.stroke();
          });
      });
    },

    drawLine(): void {
      this.ctx.beginPath();
      this.ctx.moveTo(500, 100);
      this.getShape.forEach((point) => {
        this.ctx.lineTo(point.x * 400 + 100, -point.y * 400 + 100);
			});
			this.ctx.strokeStyle = 'red';
      this.ctx.closePath();
      this.ctx.stroke();
    },
  },
});
</script>