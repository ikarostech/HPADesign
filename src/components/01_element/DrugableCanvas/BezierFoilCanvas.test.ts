import { shallowMount } from '@vue/test-utils';
import { Vector2 } from 'three/src/math/Vector2';
import BezierFoilCanvas from './BezierFoilCanvas.vue';

describe('AIAirfoilCanvas', () => {
  const wrapper = shallowMount(BezierFoilCanvas, {
    propsData: {
      radius: 10
    },
    data() {
      return {
        topControls: {
          points: [
            {
              point: new Vector2(0.0, 0.0),
              visible: false
            },
            {
              point: new Vector2(0.0, 0.15),
              visible: true
            },
            {
              point: new Vector2(0.43, 0.15),
              visible: true
            },
            {
              point: new Vector2(0.6, 0.15),
              visible: true
            },
            {
              point: new Vector2(1.0, 0.0),
              visible: false
            },
          ]
        },
        canverControls: {
          points: [
            {
              point: new Vector2(0.0, 0.0),
              visible: false
            },
            {
              point: new Vector2(0.2, 0.05),
              visible: true
            },
            {
              point: new Vector2(0.55, 0.05),
              visible: true
            },
            {
              point: new Vector2(0.65, 0.05),
              visible: true
            },
            {
              point: new Vector2(0.65, 0.05),
              visible: false
            }
          ]
        }
      }
    }
  });
  it('snapshot test', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
  it('shapeが更新された際にupdateShapeを出して親を更新する', () => {
    expect(wrapper.emitted('updateShape')).not.toBeUndefined();
  });
  it('Bezier制御点から翼型Shapeを作成', () => {
    const actual: Vector2[] = (wrapper.vm as any).getShape;
    const expected: Vector2[] = [

      new Vector2(
        1,
        0,
      ),
      new Vector2(
        0.9875,
        0.004670218131383158,
      ),
      new Vector2(
        0.975,
        0.009310753922694218,
      ),
      new Vector2(
        0.9625,
        0.013917297574746591,
      ),
      new Vector2(
        0.95,
        0.018490066225165594,
      ),
      new Vector2(
        0.9375,
        0.023029849771391227,
      ),
      new Vector2(
        0.925,
        0.027530741769139208,
      ),
      new Vector2(
        0.9125,
        0.03199631910052203,
      ),
      new Vector2(
        0.9,
        0.036425050505050495,
      ),
      new Vector2(
        0.8875,
        0.04081174144279284,
      ),
      new Vector2(
        0.875,
        0.045159094047783434,
      ),
      new Vector2(
        0.8625,
        0.04946558959295002,
      ),
      new Vector2(
        0.85,
        0.05372800225256936,
      ),
      new Vector2(
        0.8375,
        0.057944587077449705,
      ),
      new Vector2(
        0.825,
        0.0621154791154791,
      ),
      new Vector2(
        0.8125,
        0.0662388311117621,
      ),
      new Vector2(
        0.8,
        0.07031269487750551,
      ),
      new Vector2(
        0.7875,
        0.07433465390940351,
      ),
      new Vector2(
        0.775,
        0.07830150068212824,
      ),
      new Vector2(
        0.7625,
        0.08221268255188319,
      ),
      new Vector2(
        0.75,
        0.08606580383595822,
      ),
      new Vector2(
        0.7375,
        0.08985832937826291,
      ),
      new Vector2(
        0.725,
        0.09358757424947825,
      ),
      new Vector2(
        0.7125,
        0.09725069252077562,
      ),
      new Vector2(
        0.7,
        0.10084466501240695,
      ),
      new Vector2(
        0.6875,
        0.10436628590626573,
      ),
      new Vector2(
        0.675,
        0.10781214809759422,
      ),
      new Vector2(
        0.6625,
        0.11117639349991262,
      ),
      new Vector2(
        0.65,
        0.11445604688332442,
      ),
      new Vector2(
        0.6375,
        0.11764749954865499,
      ),
      new Vector2(
        0.625,
        0.1207462823572609,
      ),
      new Vector2(
        0.6125,
        0.12374449048785083,
      ),
      new Vector2(
        0.6,
        0.1266359976998275,
      ),
      new Vector2(
        0.5875,
        0.12941824390243903,
      ),
      new Vector2(
        0.575,
        0.13207639863645476,
      ),
      new Vector2(
        0.5625,
        0.13461052093973438,
      ),
      new Vector2(
        0.55,
        0.13700231238175317,
      ),
      new Vector2(
        0.5375,
        0.13924864689326694,
      ),
      new Vector2(
        0.525,
        0.1413353214049039,
      ),
      new Vector2(
        0.5125,
        0.14324436090225567,
      ),
      new Vector2(
        0.5,
        0.14496400846859564,
      ),
      new Vector2(
        0.4875,
        0.14647520058351565,
      ),
      new Vector2(
        0.475,
        0.14775622641509434,
      ),
      new Vector2(
        0.4625,
        0.14878225175918686,
      ),
      new Vector2(
        0.45,
        0.149515466739666,
      ),
      new Vector2(
        0.4375,
        0.149927027027027,
      ),
      new Vector2(
        0.425,
        0.1499912352459974,
      ),
      new Vector2(
        0.4125,
        0.14993577817531303,
      ),
      new Vector2(
        0.4,
        0.14980696469454152,
      ),
      new Vector2(
        0.3875,
        0.14961332595053525,
      ),
      new Vector2(
        0.375,
        0.1493400074617585,
      ),
      new Vector2(
        0.3625,
        0.14899294700724358,
      ),
      new Vector2(
        0.35,
        0.14856225106000256,
      ),
      new Vector2(
        0.3375,
        0.1480443437130469,
      ),
      new Vector2(
        0.325,
        0.14743957521172202,
      ),
      new Vector2(
        0.3125,
        0.14673296613627088,
      ),
      new Vector2(
        0.3,
        0.14592508007241328,
      ),
      new Vector2(
        0.2875,
        0.1450107718647453,
      ),
      new Vector2(
        0.275,
        0.14398200965335675,
      ),
      new Vector2(
        0.2625,
        0.14282484076433122,
      ),
      new Vector2(
        0.25,
        0.1415348837209302,
      ),
      new Vector2(
        0.2375,
        0.14010262213204308,
      ),
      new Vector2(
        0.225,
        0.13851627906976743,
      ),
      new Vector2(
        0.2125,
        0.13676274121721918,
      ),
      new Vector2(
        0.2,
        0.13482736377525034,
      ),
      new Vector2(
        0.1875,
        0.13269374016436442,
      ),
      new Vector2(
        0.175,
        0.13033382164438748,
      ),
      new Vector2(
        0.1625,
        0.12773028927963698,
      ),
      new Vector2(
        0.15,
        0.12486026968927103,
      ),
      new Vector2(
        0.1375,
        0.12167225766618645,
      ),
      new Vector2(
        0.125,
        0.11814236035644425,
      ),
      new Vector2(
        0.1125,
        0.11420083540302552,
      ),
      new Vector2(
        0.1,
        0.10978374490529849,
      ),
      new Vector2(
        0.0875,
        0.10480258113979043,
      ),
      new Vector2(
        0.075,
        0.09912104230876997,
      ),
      new Vector2(
        0.0625,
        0.09256735125339777,
      ),
      new Vector2(
        0.05,
        0.08485338725985843,
      ),
      new Vector2(
        0.0375,
        0.0754996058336618,
      ),
      new Vector2(
        0.025,
        0.063609397247271,
      ),
      new Vector2(
        0.0125,
        0.046785049833887045,
      ),
      new Vector2(
        0,
        0,
      ),
      new Vector2(
        0.0125,
        -0.040703442038515544,
      ),
      new Vector2(
        0.025,
        -0.0517643436119909,
      ),
      new Vector2(
        0.0375,
        -0.05818922076948444,
      ),
      new Vector2(
        0.05,
        -0.06235856562579629,
      ),
      new Vector2(
        0.0625,
        -0.06515472442137973,
      ),
      new Vector2(
        0.075,
        -0.06703927435296887,
      ),
      new Vector2(
        0.0875,
        -0.06828578807370159,
      ),
      new Vector2(
        0.1,
        -0.06905260781709445,
      ),
      new Vector2(
        0.1125,
        -0.06947365806829309,
      ),
      new Vector2(
        0.125,
        -0.06961712863656783,
      ),
      new Vector2(
        0.1375,
        -0.0695410267404489,
      ),
      new Vector2(
        0.15,
        -0.06930782213682347,
      ),
      new Vector2(
        0.1625,
        -0.06892969697953825,
      ),
      new Vector2(
        0.175,
        -0.06845409322537681,
      ),
      new Vector2(
        0.1875,
        -0.06789508244624362,
      ),
      new Vector2(
        0.2,
        -0.06726091702689971,
      ),
      new Vector2(
        0.2125,
        -0.06658100775962365,
      ),
      new Vector2(
        0.225,
        -0.0658582145536384,
      ),
      new Vector2(
        0.2375,
        -0.06510262213204308,
      ),
      new Vector2(
        0.25,
        -0.06432950255501092,
      ),
      new Vector2(
        0.2625,
        -0.06353912647861693,
      ),
      new Vector2(
        0.275,
        -0.06273705794132865,
      ),
      new Vector2(
        0.2875,
        -0.06192389089167841,
      ),
      new Vector2(
        0.3,
        -0.06111070881492825,
      ),
      new Vector2(
        0.3125,
        -0.060303668930513046,
      ),
      new Vector2(
        0.325,
        -0.059501603711302906,
      ),
      new Vector2(
        0.3375,
        -0.0587007752483166,
      ),
      new Vector2(
        0.35,
        -0.05791311383732384,
      ),
      new Vector2(
        0.3625,
        -0.05713533919601495,
      ),
      new Vector2(
        0.375,
        -0.05636821052380522,
      ),
      new Vector2(
        0.3875,
        -0.05561891254271402,
      ),
      new Vector2(
        0.4,
        -0.0548789014534348,
      ),
      new Vector2(
        0.4125,
        -0.054160523672572236,
      ),
      new Vector2(
        0.425,
        -0.05345283338408893,
      ),
      new Vector2(
        0.4375,
        -0.05270719612771879,
      ),
      new Vector2(
        0.45,
        -0.05169368456144818,
      ),
      new Vector2(
        0.4625,
        -0.05043583666484723,
      ),
      new Vector2(
        0.475,
        -0.048960414896769736,
      ),
      new Vector2(
        0.4875,
        -0.04730543141002347,
      ),
      new Vector2(
        0.5,
        -0.04549242175272848,
      ),
      new Vector2(
        0.5125,
        -0.04354063010488918,
      ),
      new Vector2(
        0.525,
        -0.04146730110033536,
      ),
      new Vector2(
        0.5375,
        -0.0392824355345322,
      ),
      new Vector2(
        0.55,
        -0.037002312381753166,
      ),
      new Vector2(
        0.5625,
        -0.03461052093973438,
      ),
      new Vector2(
        0.575,
        -0.032076398636454756,
      ),
      new Vector2(
        0.5875,
        -0.029418243902439023,
      ),
      new Vector2(
        0.6,
        -0.026635997699827496,
      ),
      new Vector2(
        0.6125,
        -0.023744490487850822,
      ),
      new Vector2(
        0.625,
        -0.020746282357260892,
      ),
      new Vector2(
        0.6375,
        -0.01764749954865498,
      ),
      new Vector2(
        0.65,
        -0.014458903944551252,
      ),
      new Vector2(
        0.6625,
        -0.014750577094667067,
      ),
      new Vector2(
        0.675,
        -0.014957658225876289,
      ),
      new Vector2(
        0.6875,
        -0.015083122568075397,
      ),
      new Vector2(
        0.7,
        -0.015132828207744226,
      ),
      new Vector2(
        0.7125,
        -0.015110182249640516,
      ),
      new Vector2(
        0.725,
        -0.015018390511870747,
      ),
      new Vector2(
        0.7375,
        -0.014860472174183056,
      ),
      new Vector2(
        0.75,
        -0.014639273165405947,
      ),
      new Vector2(
        0.7625,
        -0.01435747841485853,
      ),
      new Vector2(
        0.775,
        -0.014017623078631206,
      ),
      new Vector2(
        0.7875,
        -0.013622102839434089,
      ),
      new Vector2(
        0.8,
        -0.01317147034106371,
      ),
      new Vector2(
        0.8125,
        -0.012668933108847896,
      ),
      new Vector2(
        0.825,
        -0.012116907646092495,
      ),
      new Vector2(
        0.8375,
        -0.011517342141590746,
      ),
      new Vector2(
        0.85,
        -0.010872083850237996,
      ),
      new Vector2(
        0.8625,
        -0.010180997724146289,
      ),
      new Vector2(
        0.875,
        -0.0094458287125073,
      ),
      new Vector2(
        0.8875,
        -0.008669802641044312,
      ),
      new Vector2(
        0.9,
        -0.007854438236829596,
      ),
      new Vector2(
        0.9125,
        -0.006997033365828728,
      ),
      new Vector2(
        0.925,
        -0.006102782567973537,
      ),
      new Vector2(
        0.9375,
        -0.005173217103753163,
      ),
      new Vector2(
        0.95,
        -0.004204760091055129,
      ),
      new Vector2(
        0.9625,
        -0.0032033179741637555,
      ),
      new Vector2(
        0.975,
        -0.002168100855638979,
      ),
      new Vector2(
        0.9875,
        -0.0010988915978555613,
      ),
      new Vector2(
        1,
        0,
      ),

    ]
    expect(actual).toEqual(expected);
  })
});
