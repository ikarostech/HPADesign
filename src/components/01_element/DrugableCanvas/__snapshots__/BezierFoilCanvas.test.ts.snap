// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BezierFoilCanvas snapshot test 1`] = `
<canvas
  height="200"
  width="600"
/>
`;
